File: b944(go_restroom)/Python/go_restroom.py
from sys import stdin, stdout

limit = int(stdin.readline())
urinal = [0 for i in range(limit + 2)] # head and tail are always 0
remain_time = [0 for i in range(limit)]
remain = limit
new = 0
while True:
    line = stdin.readline()
    if not line: break
    num, time = map(int, line.split())
    if remain:
        for i in range(1, limit + 1):
            if urinal[i]: continue
            if not urinal[i-1] and not urinal[i+1]:
                new = i
                break
        else:
            for i in range(1, limit + 1):
                if not urinal[i]:
                    new = i
                    break
        urinal[new] = num
        remain_time[new - 1] = time
        remain -= 1
    else:
        stdout.write('  Not enough\n')
        new = 0
    stdout.write('Number: ' + ' '.join(map(str, urinal[1:limit+1])) \
    + '\n  Time: ' + ' '.join(map(str, remain_time)) + '\n')
    if not time and new:
        urinal[new] = 0
        remain += 1
    for i, t in enumerate(remain_time):
        if t:
            remain_time[i] -= 1
            if not remain_time[i]:
                urinal[i + 1] = 0
                remain += 1

File: b944(go_restroom)/C++/go_restroom.cpp
#include <iostream>

int main(int argc, char const *argv[]) {
  std::ios::sync_with_stdio(false);
  std::cin.tie(0);
  int limit, num, time, curr = 0;
  bool found;
  std::cin >> limit;
  int urinal[limit + 2] = {0}, remain_time[limit] = {0};
  int remain = limit;
  while (std::cin >> num >> time) {
    found = false;
    if (remain) {
      for (int i = 1; i <= limit; i++) {
        if (urinal[i]) continue;
        if (!urinal[i-1] and !urinal[i+1]) {
          curr = i;
          found = true;
          break;
        }
      }
      if (!found) for (int i = 1; i <= limit; i++)
          if (!urinal[i]) {
            curr = i;
            found = true;
            break;
          }
      urinal[curr] = num;
      remain_time[curr - 1] = time;
      remain--;
    } else {
      std::cout << " Not enough\n";
      curr = 0;
    }
    std::cout << "Number:";
    for (int i = 1; i <= limit; i++) std::cout << ' ' << urinal[i];
    std::cout << "\n  Time:";
    for (int i = 0; i < limit; i++) {
      int & t = remain_time[i];
      std::cout << ' ' << t;
      if (t) if (!--t) {
          urinal[i + 1] = 0;
          remain++;
        }
    }
    std::cout << '\n';
    if (!time && curr) {
      urinal[curr] = 0;
      remain++;
    }
  }
  return 0;
}
